app/grandchallenge/reader_studies/tasks.py

```diff
@@ -20,38 +18,33 @@
 )
 
 
-@transaction.atomic
-def add_score(obj, answer):
-    obj.calculate_score(answer)
-    obj.save(update_fields=["score"])
-
-
-@transaction.atomic
 def add_image(obj, image):
     obj.answer_image = image
     obj.save()
     image.assign_view_perm_to_creator()
     image.update_viewer_groups_permissions()
 
 
-@acks_late_micro_short_task
-def add_scores_for_display_set(*, instance_pk, ds_pk):
-    instance = Answer.objects.get(pk=instance_pk)
-    display_set = DisplaySet.objects.get(pk=ds_pk)
-    if instance.is_ground_truth:
-        for answer in Answer.objects.filter(
-            question=instance.question,
-            is_ground_truth=False,
-            display_set=display_set,
-        ):
-            add_score(answer, instance.answer)
-    else:
-        ground_truth = Answer.objects.filter(
-            question=instance.question,
-            is_ground_truth=True,
-            display_set=display_set,
-        ).get()
-        add_score(instance, ground_truth.answer)
+@acks_late_2xlarge_task
+@transaction.atomic
+def answers_from_ground_truth(*, reader_study_pk, target_user_pk):
+    reader_study = ReaderStudy.objects.get(pk=reader_study_pk)
+    target_user = get_user_model().objects.get(pk=target_user_pk)
+
+    all_answers = Answer.objects.filter(question__reader_study=reader_study)
+    if all_answers.filter(is_ground_truth=False, creator=target_user).exists():
+        raise ValueError("User already has answers")
+
+    for answer in all_answers.filter(is_ground_truth=True).all():
+        # Simplify permissions and create new answers
+        answer._state.adding = True
+        answer.id = None
+        answer.pk = None
+        answer.is_ground_truth = False
+        answer.creator = target_user
+        answer.save(calculate_score=False)
```

Comment:
There is a race condition here - there is a check to see if the user has answers, then the answers are created. In the meantime, the user could create some and there would be duplicates. To solve that `answer.full_clean()` should be called before `answer.save(...)`. That would then make the pre-flight check redundant. Hopefully duplicates are checked in the models clean method, if not, that should be added there.

See also https://realpython.com/python-lbyl-vs-eafp/

Reply:
Saving throws an `IntegrityError` here if an answer already exists: as expected. So for clearness I've removed the earlier global check.

I thought it good practice to also add a `full_clean` here. However, **that** opened up a whole can of worms with respect to the `Answer.full_clean` being broken and that it is currently being crutched by `Answer.validate()`. A small dip into that rabbit hole convinced me very quickly that this PR is not the place and I've opted to just use `Answer.validate()`. That should mark this clearly as a place where a `full_clean` should occur.

Reply:
Urgh, yes, I remember that now. That should be refactored if possible, someday.